tests/infra/trace/models.py

```diff
@@ -0,0 +1,204 @@
+    if isinstance(v, int):
+        return int(v)
+    if isinstance(v, list):
+        return [_clean_value(x) for x in v]
+    if isinstance(v, dict):
+        return {k: _clean_value(val) for k, val in v.items()}
+    return v
+
+
+class TraceStepModel(BaseModel):
+    """
+    A single step in the execution trace.
+    Represents a function call ('op'), its inputs, and its outcome.
+    """
+
+    op: str = Field(..., description="The spec function name, e.g., 'process_slots'")
+    params: dict[str, Any] = Field(
+        default_factory=dict, description="Arguments passed to the function"
+    )
+    result: Any | None = Field(
+        None, description="The return value (context var reference or primitive)"
+    )
+    error: dict[str, str] | None = Field(
+        None, description="Error details if the operation raised an exception"
+    )
+
+    @field_validator("params", "result", mode="before")
+    @classmethod
+    def sanitize_data(cls, v: Any) -> Any:
+        return _clean_value(v)
+
+
+class TraceModel(BaseModel):
+    """
+    The root schema for the trace file.
+    Contains metadata, context, and the execution trace.
+    """
+
+    metadata: dict[str, Any] = Field(..., description="Test run metadata (fork, preset, etc.)")
+    context: ContextModel = Field(default_factory=ContextModel)
+    trace: list[TraceStepModel] = Field(default_factory=list)
+
+    # Private registry state (not serialized directly, used to build the trace)
+    _hash_to_name: dict[int, str] = PrivateAttr(default_factory=dict)
+    _name_to_obj: dict[str, Any] = PrivateAttr(default_factory=dict)
+    _artifacts: dict[str, Container] = PrivateAttr(default_factory=dict)
+
+    def register_object(self, obj: Any) -> ContextVar | None:
+        """
+        Registers an object in the trace context.
+        - If it's an SSZ object (Container), it gets a hash-based name.
+        - If it's a primitive, it returns None (passed through).
+        """
+        if obj is None:
+            return None
+
+        if not isinstance(obj, Container):
+            return None
+
+        class_name = type(obj).__name__
+        if class_name not in CLASS_NAME_MAP:
+            return None  # Unknown object type
+
+        # Generate Name (Content-Addressed)
+        obj_type = CLASS_NAME_MAP[class_name]
+        root_hex = obj.hash_tree_root().hex()
+        context_name = cast(ContextVar, f"$context.{obj_type}.{root_hex}")
+        filename = f"{obj_type}_{root_hex}.ssz"
+
+        # Update Registry
+        self._hash_to_name[root_hex] = context_name
+        self._name_to_obj[context_name] = obj
+        self._artifacts[filename] = obj
+
+        # Update the public ContextObjectsModel (for output)
+        if hasattr(self.context.objects, obj_type):
+            getattr(self.context.objects, obj_type)[root_hex] = filename
+
+        return context_name
+
+    def dump_to_dir(self, output_dir: str, config: dict[str, Any] = None) -> None:
```

Comment:
Is the normal pydantic way of saving to have this as a method of the model, or having externally in another function? We should do it the pydantic way

Reply:
Let me check... I'm more used to seeing it as a method but let's do idiomatic :+1: